src/utils.h

```diff
@@ -21,6 +22,70 @@
 
 #include "compat/compat.h"
 
+/*
+ * Macro for debug messages that should *only* be present in debug builds but
+ * which should be removed in release builds. This is typically used for
+ * debug builds for development purposes.
+ *
+ * Note that some debug messages might be relevant to deploy in release build
+ * for debugging production systems. This macro is *not* for those cases.
+ */
+#ifdef TS_DEBUG
+#define TS_DEBUG_LOG(FMT, ...) elog(DEBUG2, "%s - " FMT, __func__, ##__VA_ARGS__)
+#else
+#define TS_DEBUG_LOG(FMT, ...)
+#endif
+
+#ifdef TS_DEBUG
+
+static inline const char *
+yes_no(bool value)
+{
+	return value ? "yes" : "no";
+}
+
+/* Convert datum to string using the output function. */
+static inline const char *
+datum_as_string(Oid typid, Datum value, bool is_null)
+{
+	Oid typoutput;
+	bool typIsVarlena;
+
+	if (is_null)
+		return "<NULL>";
+
+	getTypeOutputInfo(typid, &typoutput, &typIsVarlena);
+	return OidOutputFunctionCall(typoutput, value);
+}
+
+static inline const char *
+slot_as_string(TupleTableSlot *slot)
+{
+	StringInfoData info;
+	initStringInfo(&info);
+	appendStringInfoString(&info, "{");
+	for (int i = 0; i < slot->tts_tupleDescriptor->natts; i++)
+	{
+		Form_pg_attribute att = TupleDescAttr(slot->tts_tupleDescriptor, i);
+		Oid typid = att->atttypid;
+		Datum datum = slot->tts_values[i];
+		bool isnull = slot->tts_isnull[i];
```

Comment:
Are those variables really necessary? I mean all can be direct used in `datum_as_string` function call, no?

Reply:
It's a matter of style, it makes no practical difference. Removed the variables and inlined the values.